src/lib.rs

```diff
@@ -6,18 +6,203 @@ An experimental x86_64 bootloader that works on both BIOS and UEFI systems.
+/// It can currently create MBR (BIOS), GPT (UEFI), and TFTP (UEFI) images.
+pub struct DiskImageBuilder {
+    files: Vec<DiskImageFile>,
+}
+
+impl DiskImageBuilder {
+    /// Create a new instance of DiskImageBuilder, with the specified kernel.
+    pub fn new(kernel: &Path) -> Self {
+        let mut obj = Self::empty();
+        obj.set_kernel(kernel);
+        obj
+    }
+
+    /// Create a new, empty instance of DiskImageBuilder
+    pub fn empty() -> Self {
+        Self { files: Vec::new() }
+    }
+
+    /// Add or replace a kernel to be included in the final image.
+    pub fn set_kernel(&mut self, path: &Path) -> &mut Self {
+        self.add_or_replace_file(path, KERNEL_FILE_NAME)
+    }
+
+    /// Add or replace a ramdisk to be included in the final image.
+    pub fn set_ramdisk(&mut self, path: &Path) -> &mut Self {
+        self.add_or_replace_file(path, RAMDISK_FILE_NAME)
+    }
+
+    /// Configures the runtime behavior of the bootloader.
+    pub fn set_boot_config(&mut self, boot_config: &BootConfig) -> &mut Self {
+        let json =
+            serde_json::to_string_pretty(boot_config).expect("failed to serialize BootConfig");
+        let bytes = json.as_bytes();
+        self.add_or_replace_file_byte_content(bytes, CONFIG_FILE_NAME)
+    }
+
+    /// Add or replace a file from a byte array
+    pub fn add_or_replace_file_byte_content(&mut self, data: &[u8], target: &str) -> &mut Self {
+        let temp_path = temp_dir();
+        let file_name = temp_path.join("bytes.tmp");
+        fs::create_dir_all(temp_path).expect("Failed to create temp directory");
+        let mut temp_file =
+            fs::File::create(file_name.clone()).expect("Failed to create temp file");
+        temp_file
+            .write_all(data)
+            .expect("Failed to write data to temp file");
+        temp_file
+            .sync_all()
+            .expect("Failed to flush temp file to disk");
+
+        self.add_or_replace_file(&file_name, target)
+    }
+
+    /// Add or replace arbitrary files.
+    /// NOTE: You can overwrite internal files if you choose, such as EFI/BOOT/BOOTX64.EFI
+    /// This can be useful in situations where you want to generate an image, but not use the provided bootloader.
+    pub fn add_or_replace_file(&mut self, path: &Path, target: &str) -> &mut Self {
+        self.files.insert(
+            0,
+            DiskImageFile {
+                source: path.to_path_buf(),
+                destination: target.to_string(),
+            },
+        );
+        self
+    }
+    fn create_fat_filesystem_image(
+        &self,
+        internal_files: BTreeMap<&str, &Path>,
+    ) -> anyhow::Result<NamedTempFile> {
+        let mut local_map = BTreeMap::new();
+
+        for k in internal_files {
+            local_map.insert(k.0, k.1);
+        }
+
+        for f in self.files.as_slice() {
+            local_map.insert(&f.destination, f.source.as_path());
+        }
```

Comment:
These files are only added to the disk image, but not loaded on boot right? What are your plans about this? Should we try to load these from the bootloader or does the kernel need to create its own disk driver in order to access them.

In case we want to load them in the bootloader, we might need to put additional restrictions on the target file paths. Right now the BIOS bootloader only supports loading files with short file names from the root folder.

Reply:
Yes, the files are not loaded at boot.
That would be a future PR to add support, this is just to add the files (Which the kernel can load if it desires).

Still sketching out how the files will get loaded.
Doesn't the FAT library support 8.3 names with LFN? I could look into LFN support at a future date.
Could calling it out in the docs be enough?